File: 01-programming-languages/Python/01-fundamentals/03-control-flow/loops.py
"""for, while loops, and comprehensions."""

# ── For Loop ─────────────────────────────────────────────────────────
# Iterates over any iterable (list, str, dict, range, generator)
fruits = ["apple", "banana", "cherry"]
for fruit in fruits:
    print(fruit)

# With index
for i, fruit in enumerate(fruits):
    print(f"{i}: {fruit}")

# ── Range ────────────────────────────────────────────────────────────
# range(stop), range(start, stop), range(start, stop, step)
for i in range(5):          # 0, 1, 2, 3, 4
    print(i, end=" ")

for i in range(0, 10, 2):   # 0, 2, 4, 6, 8
    print(i, end=" ")

# ── While Loop ───────────────────────────────────────────────────────
count = 0
while count < 5:
    print(count, end=" ")
    count += 1

# ── Break, Continue, Else ───────────────────────────────────────────
# break — exit loop early
# continue — skip to next iteration
# else — runs if loop completes without break

for n in range(2, 10):
    for i in range(2, n):
        if n % i == 0:
            break
    else:  # This runs if the for loop didn't break
        print(f"{n} is prime", end=" ")

# ── Nested Loops ─────────────────────────────────────────────────────
matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
flat = []
for row in matrix:
    for val in row:
        flat.append(val)

# ── Zip and Unzip ────────────────────────────────────────────────────
names = ["Alice", "Bob", "Charlie"]
scores = [95, 87, 92]
paired = list(zip(names, scores))  # [('Alice', 95), ...]

unzipped_names, unzipped_scores = zip(*paired)

# ── Loop Else ────────────────────────────────────────────────────────
# else clause runs when loop finishes normally (no break)
target = 7
for n in range(2, 10):
    if n == target:
        print(f"Found {n}")
        break
else:
    print(f"{target} not found")

# ── Iterating Dicts ─────────────────────────────────────────────────
person = {"name": "Alice", "age": 30, "city": "NYC"}

for key in person:              # keys (default)
    print(key)

for key, value in person.items():  # key-value pairs
    print(f"{key}: {value}")

for value in person.values():      # values only
    print(value)

# ── List Comprehension ───────────────────────────────────────────────
squares = [x**2 for x in range(10)]
evens = [x for x in range(20) if x % 2 == 0]
matrix_transpose = [[row[i] for row in matrix] for i in range(3)]

# ── Dict Comprehension ───────────────────────────────────────────────
word_lengths = {word: len(word) for word in ["hello", "world"]}
squares_dict = {x: x**2 for x in range(5)}

# ── Set Comprehension ────────────────────────────────────────────────
unique_lengths = {len(word) for word in ["hello", "world", "hi"]}

# ── Generator Expression ─────────────────────────────────────────────
# Uses () instead of [] — lazy evaluation
sum_of_squares = sum(x**2 for x in range(1000000))

File: 01-programming-languages/Go/01-fundamentals/03-control-flow/loops.go
package main

import "fmt"

func main() {
    fmt.Println("=== Loop Patterns ===")

    // Nested loops with labels
    fmt.Println("\nNested loops with labels:")
outer:
    for i := 0; i < 3; i++ {
        for j := 0; j < 3; j++ {
            if i == 1 && j == 1 {
                continue outer
            }
            fmt.Printf("(%d,%d) ", i, j)
        }
    }
    fmt.Println()

    // Range with underscore (ignore index)
    fmt.Println("\nIgnoring index with _:")
    nums := []int{10, 20, 30}
    for _, v := range nums {
        fmt.Printf("%d ", v)
    }
    fmt.Println()

    // Range with only index
    fmt.Println("\nOnly index:")
    for i := range nums {
        fmt.Printf("Index: %d ", i)
    }
    fmt.Println()

    // Map iteration order
    fmt.Println("\nMap iteration (order not guaranteed):")
    scores := map[string]int{"Alice": 95, "Bob": 87, "Charlie": 92}
    for name, score := range scores {
        fmt.Printf("%s: %d\n", name, score)
    }

    // String range (byte vs rune)
    fmt.Println("\nString range (byte vs rune):")
    s := "Hello, 世界"
    fmt.Print("Bytes: ")
    for i := 0; i < len(s); i++ {
        fmt.Printf("%d ", s[i])
    }
    fmt.Println()
    fmt.Print("Runes: ")
    for _, r := range s {
        fmt.Printf("%c ", r)
    }
    fmt.Println()

    // Channel range
    fmt.Println("\nChannel range:")
    ch := make(chan int, 3)
    ch <- 1
    ch <- 2
    ch <- 3
    close(ch)
    for v := range ch {
        fmt.Printf("%d ", v)
    }
    fmt.Println()
}
